<launch>

  <arg name="storage" default="sqlite3" description="specify bag file storage {sqlite3, mcap}"/>

  <arg name="ouster_ns" default="ouster"
    description="Override the default namespace of all ouster nodes"/>
  <arg name="sensor_hostname"
    description="hostname or IP in dotted decimal form of the sensor"/>
  <arg name="udp_dest" default=""
    description="hostname or IP where the sensor will send data packets"/>
  <arg name="lidar_port" default="0"
    description="port to which the sensor should send lidar data"/>
  <arg name="imu_port" default="0"
    description="port to which the sensor should send imu data"/>
  <arg name="udp_profile_lidar" default=""
    description="lidar packet profile; possible values: {
    LEGACY,
    RNG19_RFL8_SIG16_NIR16,
    RNG15_RFL8_NIR8,
    RNG19_RFL8_SIG16_NIR16_DUAL,
    FUSA_RNG15_RFL8_NIR8_DUAL
    }"/>
  <arg name="lidar_mode" default=""
    description="resolution and rate; possible values: {
    512x10,
    512x20,
    1024x10,
    1024x20,
    2048x10,
    4096x5
    }"/>
  <arg name="timestamp_mode" default=""
    description="method used to timestamp measurements; possible values: {
    TIME_FROM_INTERNAL_OSC,
    TIME_FROM_SYNC_PULSE_IN,
    TIME_FROM_PTP_1588,
    TIME_FROM_ROS_TIME
    }"/>
  <arg name="ptp_utc_tai_offset" default="-37.0"
    description="UTC/TAI offset in seconds to apply when using TIME_FROM_PTP_1588"/>
  <arg name="metadata" default=""
    description="path to write metadata file when receiving sensor data"/>
  <arg name="bag_file" default=""
    description="file name to use for the recorded bag file"/>
  <arg name="viz" default="true"
    description="whether to run a rviz"/>
  <arg name="rviz_config" default="$(find-pkg-share ouster_ros)/config/viz-reliable.rviz"
    description="optional rviz config file"/>

  <arg name="sensor_frame" default="os_sensor"
    description="sets name of choice for the sensor_frame tf frame, value can not be empty"/>
  <arg name="lidar_frame" default="os_lidar"
    description="sets name of choice for the os_lidar tf frame, value can not be empty"/>
  <arg name="imu_frame" default="os_imu"
      description="sets name of choice for the os_imu tf frame, value can not be empty"/>
  <arg name="point_cloud_frame" default=""
    description="which frame to be used when publishing PointCloud2 or LaserScan messages.
    Choose between the value of sensor_frame or lidar_frame, leaving this value empty
    would set lidar_frame to be the frame used when publishing these messages."/>
  <arg name="pub_static_tf" default="true"
    description="when this flag is set to True, the driver will broadcast the TF transforms
    for the imu/sensor/lidar frames. Prevent the driver from broadcasting TF transforms by
    setting this parameter to False."/>

  <arg name="use_system_default_qos" default="true"
    description="Use the default system QoS settings"/>

  <arg name="proc_mask" default="IMG|PCL|IMU|SCAN|TLM" description="
    use any combination of the 4 flags to enable or disable specific processors"/>

  <arg name="scan_ring" default="0" description="
    use this parameter in conjunction with the SCAN flag
    and choose a value the range [0, sensor_beams_count)"/>

  <arg name="point_type" default="original" description="point type for the generated point cloud;
   available options: {
    original,
    native,
    xyz,
    xyzi,
    o_xyzi,
    xyzir
    }"/>

  <arg name="organized" default="true"
    description="generate an organzied point cloud"/>
  <arg name="destagger" default="true"
    description="enable or disable point cloud destaggering"/>

  <arg name="min_range" default="0.0"
    description="minimum lidar range to consider (meters)"/>
  <arg name="max_range" default="100000.0"
    description="minimum lidar range to consider (meters)"/>

  <arg name="azimuth_window_start" default="0" description="azimuth window start;
    values range [0, 360000] millidegrees"/>
  <arg name="azimuth_window_end" default="360000" description="azimuth window end;
    values range [0, 360000] millidegrees"/>

  <arg name="persist_config" default="false"
    description="request the sensor to persist settings"/>

  <arg name="attempt_reconnect" default="false"
    description="attempting to reconnect to the sensor after connection loss or
    sensor powered down"/>
  <arg name="dormant_period_between_reconnects" default="1.0"
    description="wait time in seconds between reconnection attempts"/>
  <arg name="max_failed_reconnect_attempts" default="2147483647"
    description="maximum number of attempts trying to communicate with the sensor.
         Counter resets upon successful connection"/>

  <arg name="auto_start" default="true"
    description="automatically configure and activate the node"/>

  <arg name="min_scan_valid_columns_ratio" default="0.0"
    description="The minimum ratio of valid columns for processing the LidarScan [0, 1]"/>

  <arg name="v_reduction" default="1" description="vertical beam reduction;
   available options: {1, 2, 4, 8, 16}"/>

  <arg name="mask_path" default="" description="path to an image file that will be used to
   mask parts of the pointcloud}"/>

  <group>
    <push-ros-namespace namespace="$(var ouster_ns)"/>
    <node_container pkg="rclcpp_components" exec="component_container_mt" name="os_container" output="screen" namespace="">
      <composable_node pkg="ouster_ros" plugin="ouster_ros::OusterSensor" name="os_sensor">
        <param name="sensor_hostname" value="$(var sensor_hostname)"/>
        <param name="udp_dest" value="$(var udp_dest)"/>
        <param name="mtp_dest" value=""/>
        <param name="mtp_main" value="false"/>
        <param name="lidar_port" value="$(var lidar_port)"/>
        <param name="imu_port" value="$(var imu_port)"/>
        <param name="udp_profile_lidar" value="$(var udp_profile_lidar)"/>
        <param name="lidar_mode" value="$(var lidar_mode)"/>
        <param name="timestamp_mode" value="$(var timestamp_mode)"/>
        <param name="metadata" value="$(var metadata)"/>
        <param name="use_system_default_qos" value="$(var use_system_default_qos)"/>
        <param name="azimuth_window_start" value="$(var azimuth_window_start)"/>
        <param name="azimuth_window_end" value="$(var azimuth_window_end)"/>
        <param name="persist_config" value="$(var persist_config)"/>
        <param name="attempt_reconnect" value="$(var attempt_reconnect)"/>
        <param name="dormant_period_between_reconnects"
              value="$(var dormant_period_between_reconnects)"/>
        <param name="max_failed_reconnect_attempts"
              value="$(var max_failed_reconnect_attempts)"/>
        <param name="auto_start" value="$(var auto_start)"/>
      </composable_node>
      <composable_node pkg="ouster_ros" plugin="ouster_ros::OusterCloud" name="os_cloud">
        <param name="sensor_frame" value="$(var sensor_frame)"/>
        <param name="lidar_frame" value="$(var lidar_frame)"/>
        <param name="imu_frame" value="$(var imu_frame)"/>
        <param name="point_cloud_frame" value="$(var point_cloud_frame)"/>
        <param name="pub_static_tf" value="$(var pub_static_tf)"/>
        <param name="timestamp_mode" value="$(var timestamp_mode)"/>
        <param name="ptp_utc_tai_offset" value="$(var ptp_utc_tai_offset)"/>
        <param name="use_system_default_qos" value="$(var use_system_default_qos)"/>
        <param name="proc_mask" value="$(var proc_mask)"/>
        <param name="scan_ring" value="$(var scan_ring)"/>
        <param name="point_type" value="$(var point_type)"/>
        <param name="organized" value="$(var organized)"/>
        <param name="destagger" value="$(var destagger)"/>
        <param name="min_range" value="$(var min_range)"/>
        <param name="max_range" value="$(var max_range)"/>
        <param name="v_reduction" value="$(var v_reduction)"/>
        <param name="mask_path" value="$(var mask_path)"/>
        <param name="min_scan_valid_columns_ratio" value="$(var min_scan_valid_columns_ratio)"/>
      </composable_node>
      <composable_node pkg="ouster_ros" plugin="ouster_ros::OusterImage" name="os_image">
        <param name="use_system_default_qos" value="$(var use_system_default_qos)"/>
        <param name="proc_mask" value="$(var proc_mask)"/>
        <param name="mask_path" value="$(var mask_path)"/>
      </composable_node>
    </node_container>
  </group>

  <include if="$(var viz)" file="$(find-pkg-share ouster_ros)/launch/rviz.launch.xml">
    <arg name="ouster_ns" value="$(var ouster_ns)"/>
    <arg name="rviz_config" value="$(var rviz_config)"/>
  </include>

  <let name="_use_bag_file_name" value="$(eval '\'$(var bag_file)\' != \'b\'')"/>
  <!-- TODO: find out why this doesn't work when passing to record -->
  <!-- <let name="_topics_to_record" value="/imu_packets /lidar_packets"/> -->

  <executable if="$(var _use_bag_file_name)" output="screen"
    cmd="ros2 bag record
      --output $(var bag_file)
      --storage $(var storage)
      /$(var ouster_ns)/imu_packets
      /$(var ouster_ns)/lidar_packets
      /$(var ouster_ns)/metadata"/>

  <executable unless="$(var _use_bag_file_name)" output="screen"
    cmd="ros2 bag record
      --storage $(var storage)
      /$(var ouster_ns)/imu_packets
      /$(var ouster_ns)/lidar_packets
      /$(var ouster_ns)/metadata"/>

</launch>
